cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/ConnectBuildOperator.java

```diff
@@ -0,0 +1,372 @@
+    private final ServiceAccountOperator serviceAccountOperations;
+    private final BuildConfigOperator buildConfigOperator;
+    private final BuildOperator buildOperator;
+
+    private final ImagePullPolicy imagePullPolicy;
+    private final List<LocalObjectReference> imagePullSecrets;
+    private final long connectBuildTimeoutMs;
+    private final PlatformFeaturesAvailability pfa;
+
+    private Reconciliation reconciliation;
+    private String namespace;
+    private KafkaConnectBuild connectBuild;
+
+    public ConnectBuildOperator(PlatformFeaturesAvailability pfa, ResourceOperatorSupplier supplier, ClusterOperatorConfig config) {
+        this.deploymentOperations = supplier.deploymentOperations;
+        this.podOperator = supplier.podOperations;
+        this.configMapOperations = supplier.configMapOperations;
+        this.serviceAccountOperations = supplier.serviceAccountOperations;
+        this.buildConfigOperator = supplier.buildConfigOperations;
+        this.buildOperator = supplier.buildOperations;
+
+        this.imagePullPolicy = config.getImagePullPolicy();
+        this.imagePullSecrets = config.getImagePullSecrets();
+        this.connectBuildTimeoutMs = config.getConnectBuildTimeoutMs();
+        this.pfa = pfa;
+    }
+
+    /**
+     * Builds a new container image with connectors on Kubernetes using Kaniko or on OpenShift using BuildConfig
+     *
+     * @param reconciliation    The reconciliation
+     * @param namespace         Namespace of the Connect cluster
+     * @param connectName       Name of the Connect cluster
+     * @param connectBuild      KafkaConnectBuild object
+     * @return                  Future for tracking the asynchronous result of the Kubernetes image build
+     */
+    public Future<BuildInfo> build(Reconciliation reconciliation, String namespace, String connectName, KafkaConnectBuild connectBuild) {
+        if (connectBuild.getBuild() == null) {
+            // Build is not configured => we should delete resources
+            return configMapOperations.reconcile(reconciliation, namespace, KafkaConnectResources.dockerFileConfigMapName(connectBuild.getCluster()), null)
+                    .compose(ignore -> podOperator.reconcile(reconciliation, namespace, KafkaConnectResources.buildPodName(connectBuild.getCluster()), null))
+                    .compose(ignore -> serviceAccountOperations.reconcile(reconciliation, namespace, KafkaConnectResources.buildServiceAccountName(connectBuild.getCluster()), null))
+                    .compose(ignore -> pfa.supportsS2I() ? buildConfigOperator.reconcile(reconciliation, namespace, KafkaConnectResources.buildConfigName(connectBuild.getCluster()), null) : Future.succeededFuture())
+                    .map(i -> new BuildInfo());
+        } else {
+            // Build exists => let's build
+            this.reconciliation = reconciliation;
+            this.namespace = namespace;
+            this.connectBuild = connectBuild;
```

Comment:
Sorry, hate to come up with more and more stuff ... I think this will not work. The `ConnectAssemblyOperator` is a long lived object shared between reconciliation. And inside it you create an instance of `ConnectBuildOperator` as a field in that object, so it will also exist for the whole operator lifecycle. Strimzi ensures isolation that two reconciliations for a single resource will not run in parallel. But there might be multiple connect clusters reconciled in parallel. So using object fields means they would overwrite them to each other. You would probably need to pass them around through the methods or have everything in single method and not in two `build(...)` methods.